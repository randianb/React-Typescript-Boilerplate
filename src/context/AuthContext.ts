import React from 'react';

export interface AuthContextInterface {
  checkingSession: boolean;
  token: string | null;
  idToken: string | null;
  expiresAt: number | null;
  isAuthenticated: boolean;
  handleAuthentication: () => void;
  handleLogin: () => void;
  handleLogout: () => void;
}

export const authContextDefaults: AuthContextInterface = {
  checkingSession: false,
  expiresAt: null,
  token: null,
  idToken: null,
  isAuthenticated: false,
  handleAuthentication: () => null,
  handleLogin: () => null,
  handleLogout: () => null
};

export const AuthContext = React.createContext<AuthContextInterface>(
  authContextDefaults
);
