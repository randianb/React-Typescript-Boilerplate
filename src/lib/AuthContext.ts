import React from 'react';

export interface AuthProfile {
  at_hash?: string;
  aud?: string;
  exp?: number;
  iat?: number;
  iss?: string;
  nonce?: string;
  sub?: string;
  locale?: string;
  name?: string;
  picture?: string;
  email?: string;
}

export interface AuthContextInterface {
  checkingSession: boolean;
  token: string | null;
  idToken: string | null;
  expiresAt: number | null;
  isAuthenticated: boolean;
  handleAuthentication: () => void;
  handleLogin: () => void;
  handleLogout: () => void;
}

export const authContextDefaults: AuthContextInterface = {
  checkingSession: false,
  expiresAt: null,
  token: null,
  idToken: null,
  isAuthenticated: false,
  handleAuthentication: () => null,
  handleLogin: () => null,
  handleLogout: () => null
};

export const AuthContext = React.createContext<AuthContextInterface>(
  authContextDefaults
);
