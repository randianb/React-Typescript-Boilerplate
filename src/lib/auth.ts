import auth0 from 'auth0-js';

export interface AuthProfile {
  at_hash?: string;
  aud?: string;
  exp?: number;
  iat?: number;
  iss?: string;
  nonce?: string;
  sub?: string;
  locale?: string;
  name?: string;
  picture?: string;
  email?: string;
}

export class Auth {
  private accessToken: string | null = null;
  private idToken: string | null = null;
  private expiresAt: number = 0;
  private profile: AuthProfile | null = null;
  private auth0 = new auth0.WebAuth({
    domain: process.env.REACT_APP_AUTH_DOMAIN || '',
    clientID: process.env.REACT_APP_AUTH_CLIENT_ID || '',
    redirectUri:
      process.env.REACT_APP_AUTH_CALLBACK || 'http://localhost:3000/callback',
    responseType: 'token id_token',
    scope: 'openid profile email'
  });

  constructor() {
    this.login = this.login.bind(this);
    this.logout = this.logout.bind(this);
    this.handleAuthentication = this.handleAuthentication.bind(this);
    this.isAuthenticated = this.isAuthenticated.bind(this);
    this.getAccessToken = this.getAccessToken.bind(this);
    this.getIdToken = this.getIdToken.bind(this);
    this.silentAuth = this.silentAuth.bind(this);
    this.getProfile = this.getProfile.bind(this);
  }

  public login() {
    this.auth0.authorize();
  }

  public handleAuthentication(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.auth0.parseHash((err, authResult) => {
        if (err) {
          return reject(err);
        }
        if (!authResult || !authResult.idToken) {
          return reject(err);
        }
        this.setSession(authResult);
        resolve();
      });
    });
  }

  public getAccessToken() {
    return this.accessToken;
  }

  public getProfile() {
    return this.profile;
  }

  public getIdToken() {
    return this.idToken;
  }

  public silentAuth(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.auth0.checkSession({}, (err, authResult) => {
        if (err) {
          return reject(err);
        }
        this.setSession(authResult);
        resolve();
      });
    });
  }

  public logout() {
    this.accessToken = null;
    this.idToken = null;
    this.expiresAt = 0;
    this.profile = null;

    this.auth0.logout({
      returnTo: process.env.REACT_APP_AUTH_RETURN_URL,
      clientID: process.env.REACT_APP_AUTH_CLIENT_ID
    });
  }

  public isAuthenticated() {
    return new Date().getTime() < this.expiresAt;
  }

  private async setSession(authResult: any) {
    const expiresAt = authResult.expiresIn * 1000 + new Date().getTime();
    this.accessToken = authResult.accessToken;
    this.idToken = authResult.idToken;
    this.profile = authResult.idTokenPayload;
    this.expiresAt = expiresAt;
  }
}
