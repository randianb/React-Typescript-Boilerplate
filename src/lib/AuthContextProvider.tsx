import React from 'react';
import { AuthContext, AuthContextInterface } from './AuthContext';
import { withRouter, RouteComponentProps } from 'react-router';
import auth0 from 'auth0-js';
class AuthContextProvider extends React.Component<RouteComponentProps> {
  private auth0: auth0.WebAuth;

  public state: AuthContextInterface = {
    token: null,
    idToken: null,
    expiresAt: 0,
    checkingSession: true,
    isAuthenticated: false,
    handleAuthentication: this.handleAuthentication.bind(this),
    handleLogin: this.handleLogin.bind(this),
    handleLogout: this.handleLogout.bind(this)
  };

  constructor(props: RouteComponentProps) {
    super(props);
    this.auth0 = new auth0.WebAuth({
      domain: process.env.REACT_APP_AUTH_DOMAIN || '',
      clientID: process.env.REACT_APP_AUTH_CLIENT_ID || '',
      redirectUri: process.env.REACT_APP_AUTH_CALLBACK,
      responseType: 'token id_token',
      scope: 'openid'
    });
  }

  public handleLogin() {
    this.auth0.authorize();
  }

  public handleAuthentication() {
    this.auth0.parseHash((err, authResult) => {
      if (authResult && authResult.accessToken && authResult.idToken) {
        this.setSession(authResult);
      } else if (err) {
        console.log('handleAuthentication error', err);
      }
      this.props.history.push('/');
    });
  }

  public getToken() {
    return this.state.token;
  }

  public getIdToken() {
    return this.state.idToken;
  }

  public setSession(authResult: auth0.Auth0DecodedHash) {
    localStorage.setItem('isLoggedIn', 'true');
    const expiresAt = authResult.expiresIn
      ? authResult.expiresIn * 1000 + new Date().getTime()
      : 0;
    this.setState({
      token: authResult.accessToken || null,
      idToken: authResult.idToken || null,
      expiresAt,
      isAuthenticated: true,
      checkingSession: false
    });
  }

  public renewSession() {
    this.auth0.checkSession({}, (err, authResult) => {
      if (authResult && authResult.accessToken && authResult.idToken) {
        this.setSession(authResult);
      } else if (err) {
        console.log('renewSession error', err);
      }
      this.setState({
        checkingSession: false
      });
    });
  }

  public handleLogout() {
    this.setState({
      token: null,
      idToken: null,
      expiresAt: 0,
      isAuthenticated: false
    });

    localStorage.removeItem('isLoggedIn');

    this.auth0.logout({
      returnTo: window.location.origin
    });

    this.props.history.push('/');
  }

  public isAuthenticated() {
    const expiresAt = this.state.expiresAt || 0;
    return new Date().getTime() < expiresAt;
  }

  public componentDidMount() {
    this.renewSession();
  }

  public render() {
    return (
      <AuthContext.Provider value={this.state}>
        {this.props.children}
      </AuthContext.Provider>
    );
  }
}

export const AppContextProviderWithRoutes = withRouter(AuthContextProvider);
