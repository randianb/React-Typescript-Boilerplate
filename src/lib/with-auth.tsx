import React from 'react';
import { AuthContext, AuthContextInterface } from './AuthContext';

export interface AuthProps {
  auth: AuthContextInterface;
}

export function withAuth<P extends AuthProps>(Component: React.ComponentType<P>) {
  return function AuthedComponent(props: Pick<P, Exclude<keyof P, keyof AuthProps>>) {
    return (
      <AuthContext.Consumer>
        {auth => <Component {...props as P} auth={auth} />}
      </AuthContext.Consumer>
    );
  };
}
