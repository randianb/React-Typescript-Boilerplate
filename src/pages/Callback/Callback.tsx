import React from 'react';
import { RouteComponentProps, withRouter } from 'react-router';
import { AuthContext } from '../../lib/AuthContext';
import { Spinner } from '../../components/Spinner/Spinner';

interface CallbackState {
  error: string | null;
}

interface CallbackProps extends RouteComponentProps<any> {}

class Callback extends React.Component<CallbackProps, CallbackState> {
  public static contextType = AuthContext;

  public state = {
    error: null
  };

  public componentDidMount() {
    try {
      this.context.handleAuthentication().then(() => {
        this.props.history.replace('/');
      });
    } catch (ex) {
      this.setState({
        error: ex.errorDescription || ex.message
      });
    }
  }

  public render() {
    return (
      <div className="callbackPage">
        {this.state.error ? (
          <div className="error">{this.state.error}</div>
        ) : (
          <Spinner text="Loading Your Profile" />
        )}
      </div>
    );
  }
}

const CallbackWithRouter = withRouter(Callback);
delete CallbackWithRouter.contextType;

export default CallbackWithRouter;
