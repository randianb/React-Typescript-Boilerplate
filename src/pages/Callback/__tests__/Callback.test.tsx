import React from 'react';
import Callback from '../Callback';
import { render, cleanup } from 'react-testing-library';
import { MemoryRouter } from 'react-router-dom';
import { authContextDefaults, AuthContext } from '../../../lib/AuthContext';

afterEach(cleanup);

const auth = authContextDefaults;
const TestComponent = (props: any) => {
  return (
    <AuthContext.Provider value={props.auth}>
      <MemoryRouter>
        <Callback />
      </MemoryRouter>
    </AuthContext.Provider>
  );
};

describe('Callback Page', () => {
  it('Renders spinner and calls handleAuthentication', () => {
    let called = false;
    auth.handleAuthentication = () => {
      called = true;
      return Promise.resolve();
    };
    const { container } = render(<TestComponent auth={auth} />);
    expect(called).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Renders error', () => {
    let called = false;
    auth.handleAuthentication = () => {
      called = true;
      throw new Error('error');
    };
    const { container } = render(<TestComponent auth={auth} />);
    expect(called).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });
});
