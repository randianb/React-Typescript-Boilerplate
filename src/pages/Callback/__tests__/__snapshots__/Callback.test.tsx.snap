// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Callback Page Renders error 1`] = `
<div
  class="callbackPage"
>
  <div
    class="error"
  >
    error
  </div>
</div>
`;

exports[`Callback Page Renders spinner and calls handleAuthentication 1`] = `
<div
  class="callbackPage"
>
  <div
    class="Spinner"
  >
    <div
      class="Spinner-spinner"
    >
      <svg
        viewBox="0 0 64 64"
      >
        <circle
          r="26"
          transform="translate(32,32)"
        />
      </svg>
    </div>
    <div
      class="Spinner-content"
    >
      Loading Your Profile
    </div>
  </div>
</div>
`;
